README.md

# Boas vindas ao repositório do projeto Fake Racer!

## Meus comentários

Quando eu era criança, sempre me interessei por como as pessoas desenvolvedoras tinham feito os jogos que eu jogava, como elas faziam as coisas se mexerem na tela, como eram animadas e como faziam os meus apertos de botão no controle se transformarem para mim em momentos marcantes.

Daí que sempre quis fazer um pequeno jogo e agora que tenho um pouco de conhecimento em JavaScript, resolvi iniciar a construção do Fake Racer como método de aprendizagem.

Dei esse nome, porque no final das contas, muita coisa é ilusão de ótica, um dos conceitos mais difíceis de acreditar são que as curvas por exemplo, não existem em jogos 2D, é apenas algo "fake", que é tratado para parecer real aos nossos olhos.

Em construção: Feito apenas com propósitos educacionais, feito com base em uma série de vídeos do YouTube e nos seguintes links:

[Extent of the Jam](http://www.extentofthejam.com/pseudo/)<br>
[Gablaxian](https://gablaxian.com/articles/creating-a-game-with-javascript/introduction/)<br>
[CodeIncomplete](https://codeincomplete.com/articles/javascript-racer/)<br>
[YT - Franks Laboratory](https://www.youtube.com/playlist?list=PLYElE_rzEw_v8TXJ_ITSSBP_ypUKfQ7K-)<br>
[YT - Vagner Araujo](https://www.youtube.com/playlist?list=PLA-J0ex1D9S32RlZhBsiurbNCYm6IG0dO)<br>

---

## Sprites reutilizadas de jogos antigos como:

Super Monaco GP (Mega Drive)<br>
Nigel Mansell's World Championship Racing (Super Nintendo)<br>
Top Gear (Super Nintendo)<br>

---

## Instalação do projeto localmente

Após cada um dos passos, haverá um exemplo do comando a ser digitado para fazer o que está sendo pedido, caso tenha dificuldades e o exemplo não seja suficiente, não hesite em me contatar em _johnatas.henrique@gmail.com_.

1. Abra o terminal e crie um diretório no local de sua preferência com o comando **mkdir**:
```javascript
  mkdir projetos-johnatas
```

2. Entre no diretório que acabou de criar e depois clone o projeto:
```javascript
  cd projetos-johnatas
  git clone git@github.com:johnatas-henrique/meme-generator.git
```

3. Abra o diretório do projeto clonado e dê duplo clique no arquivo **index.html** e o projeto vai abrir em seu navegador, não é necessário instalar nada mais.

---

## Link para o deploy:
https://johnatas-henrique.github.io/fake-racer/


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.19.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.18.0...v2.19.0) (2022-03-12)


### Features

* add difficulty settings - thanks [@saraivais](https://github.com/saraivais) ([65433a1](https://github.com/johnatas-henrique/fake-racer/commit/65433a193304393fad2c483284abf5ea68949624))

## [2.18.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.17.0...v2.18.0) (2022-02-15)


### Features

* adds new track - Canada ([a93af8e](https://github.com/johnatas-henrique/fake-racer/commit/a93af8eda59f0e11ef43e2515381f5e626c79426))


### Bug Fixes

* fixes old variable names ([fa20523](https://github.com/johnatas-henrique/fake-racer/commit/fa205237b4d3967922fa2b4eaabfc00f293f6328))

## [2.17.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.15.0...v2.17.0) (2022-02-07)


### Features

* adds control buttons on race screen ([0a74beb](https://github.com/johnatas-henrique/fake-racer/commit/0a74beb2397b92164f452b2ebe4dcb50ffe0de4b))
* adds fullscreen button ([83b69d6](https://github.com/johnatas-henrique/fake-racer/commit/83b69d6e61340c2c00cc8f2b75a703c5ab2fc09e))


### Bug Fixes

* fixes OK button malfunctioning ([a1f2910](https://github.com/johnatas-henrique/fake-racer/commit/a1f2910cb23a34de0c6618a0a079c2536631a82f))

## [2.16.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.15.0...v2.16.0) (2022-01-30)


### Features

* adds analogic speedometer ([10ed5ad](https://github.com/johnatas-henrique/fake-racer/commit/10ed5adb51253f910095ceaf6177afcca4be60b9))
* adds menu particles effect ([b3b51c7](https://github.com/johnatas-henrique/fake-racer/commit/b3b51c7bd0aa5fa7714bb14a235864ca3cc27d9d))
* adds random rain on races ([014ae78](https://github.com/johnatas-henrique/fake-racer/commit/014ae7862bc5409f57773e6e82c6850e5a53c17c))

## [2.15.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.14.0...v2.15.0) (2022-01-30)


### Features

* adds car's segment list ([ed79749](https://github.com/johnatas-henrique/fake-racer/commit/ed797499e8cdb5bfb4a12ea989490e98575c126c))
* adds collision between opponents ([7283270](https://github.com/johnatas-henrique/fake-racer/commit/7283270cfd6b8d66ce5ab46405d52e619a4a19f7))
* adds mrdoob's FPS counter ([17345e8](https://github.com/johnatas-henrique/fake-racer/commit/17345e8ed6bc8d4723310309ffcc9a26cc8d1d6d))
* adds player collision detection ([e7cf117](https://github.com/johnatas-henrique/fake-racer/commit/e7cf117e4dc73a046e1c6b2e61d6680b87fb7110))


### Bug Fixes

* fixes parallax background side scrolling - thanks @ThiagoCarreiraVallim ([e0dc91d](https://github.com/johnatas-henrique/fake-racer/commit/e0dc91da7472908da1443538cc622b775b2c1b5e))
* reposition opponent names and shows FPS only in race screen ([b5c5615](https://github.com/johnatas-henrique/fake-racer/commit/b5c5615a0b58165e4d11538b2c0c450813ecaac8))

## [2.14.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.13.0...v2.14.0) (2022-01-27)


### Features

* adds checkered flag on track start ([7f75984](https://github.com/johnatas-henrique/fake-racer/commit/7f75984aed24a8347cb0048a52829b39b7351f97))
* adds opponent name over the opponent car ([989e270](https://github.com/johnatas-henrique/fake-racer/commit/989e270b73e9a03de046cff0f17ad011316d36af))
* adds speed on Sprite class ([663ba94](https://github.com/johnatas-henrique/fake-racer/commit/663ba94a53965fb82e2704523caa770c12981c06))


### Bug Fixes

* fixes double confirmation on menu screen ([3a91b11](https://github.com/johnatas-henrique/fake-racer/commit/3a91b111bea6d8d25b674f749c5276a64fc0d5e1))

## [2.13.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.12.0...v2.13.0) (2022-01-27)


### Features

* adds pause function on race screen ([94d2ab7](https://github.com/johnatas-henrique/fake-racer/commit/94d2ab77490f159ef09125e26a713d41e50e36fa))
* adds race start phase ([9ed06ce](https://github.com/johnatas-henrique/fake-racer/commit/9ed06cef3e914a79b7d2c54754bb34c3ba84472d))

## [2.12.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.11.0...v2.12.0) (2022-01-12)


### Features

* adds music on exiting title screen ([9f9c530](https://github.com/johnatas-henrique/fake-racer/commit/9f9c530042d017ec94c3f430e1774d67e7493cd3))
* creates rectangle with rounded corners method ([603174c](https://github.com/johnatas-henrique/fake-racer/commit/603174ca5ff4c2070f2bad2e4db5838c1d701c51))
* improves title and menu screen ([aeb4f53](https://github.com/johnatas-henrique/fake-racer/commit/aeb4f5380542d5a25eb43ffd96098d380984924a))

## [2.11.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.10.0...v2.11.0) (2022-01-12)


### Features

* adds Monaco track ([c6df22f](https://github.com/johnatas-henrique/fake-racer/commit/c6df22fdba5c4b146ff777051b913c321e54ffa2))
* adds tunnel creation method ([770c340](https://github.com/johnatas-henrique/fake-racer/commit/770c340ba1e020b55449c39b90de4aa4836d1fe0))

## [2.10.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.9.0...v2.10.0) (2021-12-22)


### Features

* adds player's position on screen ([41bee7c](https://github.com/johnatas-henrique/fake-racer/commit/41bee7ccfb2a67082c63686b168b35738f7c49ad))
* adds track position method to all cars ([a63dfbe](https://github.com/johnatas-henrique/fake-racer/commit/a63dfbe555a44bb4d0b7cd085dc66f1ff31b2f43))


### Bug Fixes

* fixes laptime bug ([ed0996c](https://github.com/johnatas-henrique/fake-racer/commit/ed0996caa555fa457340d370e158f23a63d45230))

## [2.9.0](https://github.com/johnatas-henrique/fake-racer/compare/v2.8.0...v2.9.0) (2021-12-22)


### Features

* adds menu screen ([2af28b6](https://github.com/johnatas-henrique/fake-racer/commit/2af28b69ed6fb9549031c18615c198c5eae2f75b))

js/menu.js
import {
  handleInput, tracks, startPosition, drivers, resource,
} from './util.js';
import Opponent from './opponent.js';
import Sprite from './sprite.js';

class Menu {
  constructor(width, height, animations) {
    this.showMenu = 0;
    this.height = height;
    this.state = 'title';
    this.width = width;
    this.menuY = 0;
    this.menuX = 5;
    this.updateTime = 6 / 60;
    this.updateAnimationsTime = 1 / 60;
    this.menuPhrase = {
      0: 'Circuito: ',
      1: 'Oponentes: ',
      2: 'Dificuldade: ',
      3: 'Música: ',
      4: 'Volume da música: ',
      5: 'Iniciar ',
    };
    this.menu = {
      0: Object.keys(tracks),
      1: ['1', '3', '5', '7', '9', '11', '13', '15', '17', '19'],
      2: ['novato', 'corredor', 'campeão'],
      3: ['não', 'sim'],
      4: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10'],
      5: ['corrida'],
    };
    this.selectedOptions = {
      0: 'canada',
      1: '19',
      2: 'novato',
      3: 'não',
      4: '1',
      5: 'corrida',
    };
    this.arrowUpBlink = 0;
    this.arrowDownBlink = 0;
    this.menuTitle = { pos: 0, direction: 1 };
    this.animations = animations;
  }

  adjustDifficulty() {
    if (this.selectedOptions[2] === 'novato') return 0.87;
    if (this.selectedOptions[2] === 'corredor') return 0.935;
    return 1;
  }

  startRace(player, road, opponents, director) {
    const roadParam = road;
    const zero = 0;
    drivers.forEach((driver) => opponents.push(new Opponent(
      driver.power * this.adjustDifficulty(),
      startPosition(tracks[this.selectedOptions[zero]].trackSize, driver.position),
      driver.trackSide, 'opponents', driver.name, driver.carColor,
    )));

    opponents.forEach((opponentNumber) => opponentNumber.create());
    opponents.splice(this.selectedOptions[1], opponents.length);
    roadParam.trackName = this.selectedOptions[zero];
    roadParam.create();
    player.create(this, tracks[this.selectedOptions[zero]].trackSize);
    director.create(road, this.selectedOptions[0]);
  }

  update(player, road, opponents, director) {
    const {
      arrowup, arrowdown, arrowleft, arrowright,
    } = handleInput.map;
    const maxX = Object.keys(this.menu).length - 1;
    const maxY = this.menu[this.menuX].length - 1;
    if (handleInput.mapPress.enter && !this.showMenu) {
      this.selectedOptions[3] = 'sim';
      this.showMenu = 1;
      this.menuTitle.pos = 0;
      handleInput.mapPress.enter = false;
    }

    if (this.showMenu) {
      if (!arrowdown) this.arrowDownBlink = false;
      if (!arrowup) this.arrowUpBlink = false;

      if (this.menuX < maxX && arrowdown) {
        this.arrowDownBlink = !this.arrowDownBlink;
        this.menuX += 1;
        this.menuY = this.menu[this.menuX]
          .findIndex((item) => item === this.selectedOptions[this.menuX]);
      } else if (this.menuX === maxX && arrowdown) {
        this.arrowDownBlink = 1;
        this.menuX = 0;
        this.menuY = this.menu[this.menuX]
          .findIndex((item) => item === this.selectedOptions[this.menuX]);
      }

      if (this.menuX > 0 && arrowup) {
        this.arrowUpBlink = 1;
        this.menuX -= 1;
        this.menuY = this.menu[this.menuX]
          .findIndex((item) => item === this.selectedOptions[this.menuX]);
      } else if (this.menuX === 0 && arrowup) {
        this.arrowUpBlink = 1;
        this.menuX = maxX;
        this.menuY = this.menu[this.menuX]
          .findIndex((item) => item === this.selectedOptions[this.menuX]);
      }

      if (this.menuY < maxY && arrowright) this.menuY += 1;
      else if (this.menuY === maxY && arrowright) this.menuY = 0;

      if (this.menuY > 0 && arrowleft) this.menuY -= 1;
      else if (this.menuY === 0 && arrowleft) this.menuY = maxY;

      const lastMenuOption = Object.keys(this.menu).length - 1;

      if (this.menuX !== lastMenuOption) {
        this.selectedOptions[this.menuX] = this.menu[this.menuX][this.menuY];
        handleInput.mapPress.enter = false;
      }

      if (handleInput.mapPress.enter && this.menuX === lastMenuOption) {
        const pauseBtn = document.querySelector('#pauseBtn');
        const fps = document.querySelector('#fps');
        const mute = document.querySelector('#mute');
        pauseBtn.classList.toggle('hidden');
        mute.classList.toggle('hidden');
        const okBtn = document.querySelector('.rightControls').firstElementChild;
        okBtn.classList.toggle('hidden');
        this.startRace(player, road, opponents, director);
        this.state = 'race';
        handleInput.mapPress.enter = false;
        fps.firstElementChild.classList.remove('hidden');
      }
    }
  }

  static drawButtons(render, x, y, size, text) {
    render.drawCircle(x, y + 3, size, 0, Math.PI * 2);
    render.drawText('black', text, x, y, 1.3, 'OutriderCond', 'center');
  }

  render(render) {
    this.animations.forEach((item) => item.render(render));
    render.drawText('#EB4844', 'Fake Racer', 320, 30, 4, 'OutriderCondBold');

    if (!this.showMenu) {
      if (this.menuTitle.pos >= 12) this.menuTitle.direction = -1;
      if (this.menuTitle.pos <= -12) this.menuTitle.direction = 1;
      this.menuTitle.pos += (this.menuTitle.direction / 2);
      if (window.navigator.maxTouchPoints) {
        render.drawText('black', 'Aperte OK para iniciar', 320, 180 + this.menuTitle.pos);
      } else {
        render.drawText('black', 'Aperte ENTER para iniciar', 320, 180 + this.menuTitle.pos);
      }
    }

    if (this.showMenu) {
      if (this.menuTitle.pos >= 4) this.menuTitle.direction = -1;
      if (this.menuTitle.pos <= -4) this.menuTitle.direction = 1;
      this.menuTitle.pos += (this.menuTitle.direction / 2);
      const maxX = Object.keys(this.menu).length - 1;
      const menuLow = this.menuX - 1 >= 0 ? this.menuX - 1 : maxX;
      const menuHigh = this.menuX + 1 <= maxX ? this.menuX + 1 : 0;
      const lowText = `${this.menuPhrase[menuLow]} ${this.selectedOptions[menuLow].toLocaleUpperCase()}`;
      const highText = `${this.menuPhrase[menuHigh]} ${this.selectedOptions[menuHigh].toLocaleUpperCase()}`;

      render.roundRect('#2C69EB', 100, 100, 440, 170, 20, true, false);
      render.drawText('#FFFAF4', lowText, 320, 180 - 45, 1.6);
      console.log('1', this.menuPhrase[this.menuX], '2', this.menu[this.menuX][this.menuY]);
      const phrase = `${this.menuPhrase[this.menuX]} ${this.menu[this.menuX][this.menuY].toLocaleUpperCase()}`;
      render.drawText('#050B1A', phrase, 320, 180 + (this.menuTitle.pos / 4), 1.6);
      render.drawText('#FFFAF4', highText, 320, 180 + 45, 1.6);

      if (window.navigator.maxTouchPoints) {
        Menu.drawButtons(render, 145, 310, 15, 'U');
        Menu.drawButtons(render, 185, 310, 15, 'D');
        Menu.drawButtons(render, 225, 310, 15, 'L');
        Menu.drawButtons(render, 265, 310, 15, 'R');
        render.drawText('black', 'Navegar', 150, 345, 1.3, 'OutriderCond', 'left');
        Menu.drawButtons(render, 418, 310, 18, 'OK');
        render.drawText('black', 'Confirmar', 490, 345, 1.3, 'OutriderCond', 'right');
      } else {
        const arrowKeys = new Sprite();
        arrowKeys.image = resource.get('arrowKeys');
        arrowKeys.name = 'mnArrowKeys';

        const enterKey = new Sprite();
        enterKey.image = resource.get('enterKey');
        enterKey.name = 'mnEnterKey';

        render.drawText('black', 'Navegar', 590, 320, 1.3, 'OutriderCond', 'right');
        render.renderingContext.drawImage(arrowKeys.image, 595, 310, 28, 18);
        render.drawText('black', 'Confirmar', 590, 345, 1.3, 'OutriderCond', 'right');
        render.renderingContext.drawImage(enterKey.image, 597, 335, 23, 18);
      }

      if (this.arrowUpBlink) {
        render.drawText('#050B1A', 'c', 520, 140, 2, 'Arrows');
      } else {
        render.drawText('#FFFAF4', 'c', 520, 140, 2, 'Arrows');
      }
      if (this.arrowDownBlink) {
        render.drawText('#050B1A', 'd', 520, 240, 2, 'Arrows');
      } else {
        render.drawText('#FFFAF4', 'd', 520, 240, 2, 'Arrows');
      }
    }
  }
}

export default Menu;
